first draft of verification design doc (#470)
google/exposure-notifications-server: docs/design/verification_protocol.md

The actual process of issuing these certificates is not covered in this
document, but will be at a later date.

## Motivation

The current exposure notifications publication protocol only has the ability
to authenticate that a request came from a mobile device via the vendor
specific device attestation protocols. This doesn't give us the ability
to certify that the diagnosis came from a trusted (by the server) public
health authority in the jurisdiction.

## Definitions

* PHA : Public Health Authority
* JWT : JSON Web Token
* ECDSA : Elliptic Curve Digital Signature Algorithm
* PIN : personal identification number. Could be a different authenticator.
* HMAC : hash-based message authentication code
* TEK : Temporary Exposure Key

## High Level Flow

1. App user is diagnosed with Covid-19. PHA issues a PIN code to the user.
2. The user enters the PIN code on their App
3. The app creates a HMAC value that combines the TEKs from the device along
   with calculated transmission risk value, rolling period start and rolling
	 period count values (the key data + metadata).
	 These values are sent to the PHA server that issued the PIN.

> 4. If the PIN is valid, the PHA issues a JWT that is signed using their

Reviewer: Can you talk more about PIN generation and validation? Or is that out of scope here?

Author: out of scope, will be in the PHA server design

   ECDSA p-256 algorithm. The JWT includes additional claims about the data
	 (see below).
5. The app on the user's device sends this signed JWT to the exposure
   notifications server (this project).

> 6. If the JWT is valid and signed by a trusted PHA (verified because the public

Reviewer: Maybe emphasize this is an offline verification?

> 6. If the JWT is valid and signed by a trusted PHA (verified because the public

Reviewer: We'll also somehow need to unpack with PHA key version did the signing (in case of rotation).

Author: added a `keyVersion` claim

	 key has been previously shared with the server), then the keys are imported
	 into the server for distribution to other devices in the geography.

> 

Reviewer: Should we store any of this verification alongside the data?

Author: That's a good question.

We could store the `iss` field, as in which authority issued this diagnosis. 

We could also build in storage of the `phadata` - but as the reference sever won't make use of it, I lean against storing it. 

Good debate to be had.

Reviewer: I don't like the idea of the TEK server being used to "lookup" a verification, but it could be useful in the opposite direction (given a TEK, from where was it verified, when, etc)

Author: I'm inclined to agree that we shouldn't sore this in the reference

# Public Health Authority Diagnosis Verification Protocol

This design covers the exposure notifications server's ability to verify
diagnosis certifications from public health authorities.

The actual process of issuing these certificates is not covered in this
document, but will be at a later date.

## Motivation

The current exposure notifications publication protocol only has the ability
to authenticate that a request came from a mobile device via the vendor

> specific device attestation protocols. This doesn't give us the ability

Reviewer 2: ```suggestion
specific device attestation protocols. This doesn't provide a way
```

2. The user enters the PIN code on their App
3. The app creates a HMAC value that combines the TEKs from the device along
   with calculated transmission risk value, rolling period start and rolling
	 period count values (the key data + metadata).
	 These values are sent to the PHA server that issued the PIN.
4. If the PIN is valid, the PHA issues a JWT that is signed using ECDSA over
   the P-256 elliptic curve with SHA-256 as a hash function. The JWT includes
   additional claims about the data (see below).
5. The app on the user's device sends this signed JWT to the exposure
   notifications server (this project).
6. If the JWT is valid and signed by a trusted PHA (verified because the public
	 key has been previously shared with the server), then the keys are imported
	 into the server for distribution to other devices in the geography. Because
   we are using pre-shared public keys, this verification can be done with
   offline verification.

## JWT Verification + Accepted Claims

The exposure notifications server will accept this JWT in the current
`verificationPayload` field in the exposure reporting API. When this is present,

> device attestations should be disabled. In addition, the exposure publishing

Reviewer 2: Is 'should' correct? Or is it 'may'? If may be helpful to expand why? I agree both are not necessary.

Author: I think should is correct. I also think we should move the device verification into the reference implementation of the verification server.

verification. This is to allow for key rotation.
* `aud` : The audience must be as configured for this installation of the
exposure notifications server.
* `iat` : The unix timestamp at which the token was issued.
* `exp` : The unix timestamp at which the token will expire.
* `nbf` : If present, the "not before" timestamp will be honored.

We also prescribe a set of private claims to transmit data from the PHA
verification server to the exposure notifications server.

* `tekhmac` : The HMAC of the TEKs that was presented to the PHA verification
server. This must be calculated in a specific way (see below).
* `phadata` : Contains a map of claims for the PHA to communicate back
to the mobile app and/or the exposure notifications server. This specific
implementation of the exposure notifications server will disregard and never
save the `phadata`
* `trisk` : Contains an array of transmission risk overrides to enact when
importing the associated keys. If data is present in this field, it will
override the data in the upload from the device.
* `keyVersion` : The version of the public key to use for verification.

### Transmission Risk Overrides

> An array of JSON objects of the following structure:

Reviewer 2: would it be useful to require these be ordered? or should we specify sorting on the `sinceRollingPeriod` before computation

Author: we can suggest it - the server will sort upon receipt either way
